src/test/java/org/cryptomator/cryptolib/v3/UVFIntegrationTest.java

```diff
@@ -0,0 +1,172 @@
+		Assertions.assertEquals(13, result.length);
+		Assertions.assertEquals("Hello, World!", new String(result, StandardCharsets.UTF_8));
+	}
+
+	@Test
+	@DisplayName("create reference directory structure")
+	public void testCreateReferenceDirStructure(@TempDir Path vaultDir) throws IOException {
+		DirectoryContentCryptor dirContentCryptor = cryptor.directoryContentCryptor();
+
+		// ROOT
+		DirectoryMetadata rootDirMetadata = cryptor.directoryContentCryptor().rootDirectoryMetadata();
+		String rootDirPath = dirContentCryptor.dirPath(rootDirMetadata);
+		String rootDirUvfFilePath = rootDirPath + "/dir.uvf";
+		byte[] rootDirUvfFileContents = dirContentCryptor.encryptDirectoryMetadata(rootDirMetadata);
+		Files.createDirectories(vaultDir.resolve(rootDirPath));
+		Files.write(vaultDir.resolve(rootDirUvfFilePath), rootDirUvfFileContents);
+		DirectoryContentCryptor.Encrypting filesWithinRootDir = dirContentCryptor.fileNameEncryptor(rootDirMetadata);
+
+		// ROOT/foo.txt
+		String fooFileName = filesWithinRootDir.encrypt("foo.txt");
+		String fooFilePath = rootDirPath + "/" + fooFileName;
+		byte[] fooFileContents = encryptFile(StandardCharsets.UTF_8.encode("Hello Foo"), cryptor);
+		Files.write(vaultDir.resolve(fooFilePath), fooFileContents);
+
+		// ROOT/subdir
+		DirectoryMetadata subDirMetadata = dirContentCryptor.newDirectoryMetadata();
+		String subDirName = filesWithinRootDir.encrypt("subdir");
+		String subDirUvfFilePath1 = rootDirPath + "/" + subDirName + "/dir.uvf";
+		byte[] subDirUvfFileContents1 = dirContentCryptor.encryptDirectoryMetadata(subDirMetadata);
+		Files.createDirectories(vaultDir.resolve(rootDirPath + "/" + subDirName));
+		Files.write(vaultDir.resolve(subDirUvfFilePath1), subDirUvfFileContents1);
+		String subDirPath = dirContentCryptor.dirPath(subDirMetadata);
+		String subDirUvfFilePath2 = subDirPath + "/dir.uvf";
+		byte[] subDirUvfFileContents2 = dirContentCryptor.encryptDirectoryMetadata(subDirMetadata);
+		Files.createDirectories(vaultDir.resolve(subDirPath));
+		Files.write(vaultDir.resolve(subDirUvfFilePath2), subDirUvfFileContents2);
+		DirectoryContentCryptor.Encrypting filesWithinSubDir = dirContentCryptor.fileNameEncryptor(subDirMetadata);
+
+		// ROOT/subdir/bar.txt
+		String barFileName = filesWithinSubDir.encrypt("bar.txt");
+		String barFilePath = subDirPath + "/" + barFileName;
+		byte[] barFileContents = encryptFile(StandardCharsets.UTF_8.encode("Hello Bar"), cryptor);
+		Files.write(vaultDir.resolve(barFilePath), barFileContents);
+
+		// set breakpoint here to inspect the created directory structure
+		System.out.println(vaultDir);
+
+	}
+
+	private static byte[] encryptFile(ByteBuffer cleartext, Cryptor cryptor) throws IOException {
+		try (ByteArrayOutputStream baos = new ByteArrayOutputStream();
+			 EncryptingWritableByteChannel ch = new EncryptingWritableByteChannel(Channels.newChannel(baos), cryptor)) {
+			ch.write(cleartext);
+			return baos.toByteArray();
+		}
+	}
```

Comment:
Current implentation only writes header to file. `baos.toByteArray()` happens before `encryptAndFlushBuffer()` is called upon closing the channel:
```suggestion
    private static byte[] encryptFile(ByteBuffer cleartext, Cryptor cryptor) throws IOException {
        try (ByteArrayOutputStream baos = new ByteArrayOutputStream()) {
            try (
                    EncryptingWritableByteChannel ch = new EncryptingWritableByteChannel(Channels.newChannel(baos), cryptor)) {
                ch.write(cleartext);
            }
            return baos.toByteArray();
        }
    }
```

Reply:
fixed in f2745ea, size of "Hello Word" files is now 105 bytes.